Add validation for GOAWAY and MAX_PUSH_ID frames in outbound control

comchi opened the pull request:
Hi, I'm studying how HTTP/3 works and noticed a TODO in an outbound control stream validator,  so I thought that I can help this project at least with it. I hope I understood RFC and handler purpose correctly.  

I split the changes into two commits because I'm not sure that I fixed the `Http3UnidirectionalStreamInboundHandlerTest.java:177` in the second commit in a right way

What I did: 
1. Added validation for the GOAWAY frame for the client side. 
2. Tried to implement ``` MAX_PUSH_ID frame cannot reduce the maximum Push
   ID; receipt of a MAX_PUSH_ID frame that contains a smaller value than
   previously received MUST be treated as a connection error of type
   H3_ID_ERROR``` from [this chapter](https://datatracker.ietf.org/doc/html/draft-ietf-quic-http-32#section-7.2.7)
3. Check that MAX_PUSH_ID wasn't sent by server
4. Since outbound handle works in two `modes`  (client and server) I parametrized test in the same manner as it was done in `Http3ControlStreamInboundHandlerTest`

NiteshKant commented:
Thanks a lot @comchi for the PR, I will review it in a bit but till then can you please go over the netty contribution checklist: https://netty.io/wiki/developer-guide.html#checklist

Specifically the part about the commit message and signing the CLA.

comchi commented:
@NiteshKant Hi, I used the commit message template from the checklist and signed CLA

NiteshKant commented on src/test/java/io/netty/incubator/codec/http3/Http3ControlStreamOutboundHandlerTest.java:
```diff
@@ -83,32 +97,75 @@ public void testGoAwayIdDecreaseWorks() throws Exception {
+    }
+
+    @Test
+    public void testMaxPushIdFailsWhenReduced() throws Exception {
```
Looks like this test and below can reuse some code

comchi replied:
after I reverted the code, doesn't seem like I can reuse something here

NiteshKant commented on src/test/java/io/netty/incubator/codec/http3/Http3UnidirectionalStreamInboundHandlerTest.java:
```diff
@@ -174,7 +174,7 @@ private void testPushStream(long maxPushId) throws Exception {
         }
 
-        if (maxPushId >= 0) {
+        if (!server && maxPushId >= 0) {
```
I don't think this is needed once you address my comments on the `Http3ControlStreamOutboundHandler`

comchi replied:
this is removed

comchi commented:
Hi @NiteshKant, thanks for your review.

I fixed review notes, but I have two questions:
1. Should we set GOAWAY id as new `sentMaxPushId` value when `Http3ControlStreamOutboundHandler` is not server?    
I'm not sure that overwriting `sentMaxPushId` with the value from `GOAWAY` is correct. I think that the actual limitation of maximum push id can be calculated using both values from `GOAWAY` and `MAX_PUSH_ID` frames when it is required. 
I wanted to see how restriction based on `sentMaxPushId` looks like in codebase (checked `Http3RequestStreamValidationHandler` and `Http3PushStreamClientValidationHandler`), but didn't find any. 
Also tried to check assumption with quiche implementation, but it looks like they still haven't implemented this fully https://github.com/cloudflare/quiche/blob/76920d868d0bd6b076809fde5f17f69a5a791669/src/h3/mod.rs#L1375
So, I left it as it was. Please let me know if I misunderstood something. 

2. I added validation to the outbound handler that client GOAWAY id is not greater than `max_push_id`. Should I add the validation to the inbound handler that the GOAWAY id that server receives is not greater than `receivedMaxPushId` it received before?

comchi commented on src/main/java/io/netty/incubator/codec/http3/Http3ControlStreamOutboundHandler.java:
```diff
@@ -93,34 +93,62 @@ public void channelInactive(ChannelHandlerContext ctx) {
+        }
+
+        if (server) {
+            promise.setFailure(new Http3Exception(Http3ErrorCode.H3_FRAME_UNEXPECTED, "MAX_PUSH_ID sent by server"));
```
thanks, I reverted my commit with this check

comchi commented on src/main/java/io/netty/incubator/codec/http3/Http3ControlStreamOutboundHandler.java:
```diff
@@ -93,34 +93,62 @@ public void channelInactive(ChannelHandlerContext ctx) {
+        long id = goAwayFrame.id();
+
+        // See https://tools.ietf.org/html/draft-ietf-quic-http-32#section-5.2
+        if (server && id % 4 != 0) {
```
Wasn't sure that I got it right from RFC, that's why didn't introduce it, thanks for the clarification!

comchi commented on src/main/java/io/netty/incubator/codec/http3/Http3ControlStreamOutboundHandler.java:
```diff
@@ -93,34 +93,62 @@ public void channelInactive(ChannelHandlerContext ctx) {
+        } else if (msg instanceof Http3GoAwayFrame) {
+            valid = handleHttp3GoAwayFrame(promise, (Http3GoAwayFrame) msg);
+        } else {
+            assert msg instanceof Http3UnknownFrame;
```
rewrote it as it was suggested below

NiteshKant commented:
> Should we set GOAWAY id as new sentMaxPushId value when Http3ControlStreamOutboundHandler is not server?

Nope that will be incorrect. The validation that we want to do (and what the code currently is doing in this PR) is that the ID specified in the `GO_AWAY` frame is less than the max allowed push_ID for the connection. Since, when writing, go_away is indicating the last push_id that will be processed by the client, it should be less than what the client allowed the server to send in the first place through an earlier max_push_id frame.

> Should I add the validation to the inbound handler that the GOAWAY id that server receives is not greater than receivedMaxPushId it received before?

Yeah I think that will be reasonable to do.

comchi commented:
@NiteshKant, I added new validation mentioned in the previous message to the inbound handler.  
I fixed all the comments, anything else you think I need to change, or now it is good?

NiteshKant commented on src/test/java/io/netty/incubator/codec/http3/Http3ControlStreamOutboundHandlerTest.java:
```diff
@@ -83,32 +99,91 @@ public void testGoAwayIdDecreaseWorks() throws Exception {
+    }
+
+    @Test
+    public void testClientGoAwayIdIsNotLimitedWhenMaxPushIdIsNotSet() throws Exception {
```
This test is actually interesting. It tells me that I may have misunderstood the validations that are required to be done for the ID in go_away. 

The [`Connection Shutdown` section](https://quicwg.org/base-drafts/draft-ietf-quic-http.html#connection-shutdown) in the spec says:

> An endpoint that is attempting to gracefully shut down a connection can send a GOAWAY frame with a value set to the maximum possible value (2^62-4 for servers, 2^62-1 for clients). 

Which suggests that we should perhaps not restrict the IDs that are sent/received in the go_away other than the validations specified in the spec:
- Ids are correct depending on client/server
- ID is not greater than the previous ID.

Apologies for the back-and-forth and the misdirection

comchi replied:
no problem, it was in two separate commits, so I just reverted them. Thanks for your time reviewing this

comchi commented:
@NiteshKant To me, it looks like I fixed all the comments, or did I miss something?

NiteshKant reviewed (approved):
Thanks so much @comchi
